src/engine/PathFinding.cpp

```diff
@@ -0,0 +1,137 @@
+#include "PathFinding.hpp"
+#include <set>
+
+using namespace admirals;
+
+std::deque<Vector2>
+PathFinding::FindPath(const QuadTree &quadTree, const Vector2 &start,
+                      const Vector2 &dest, const Vector2 &pathSize,
+                      const std::unordered_set<float> &checkedOrders,
+                      float detailLevel) {
+    const Rect bounds = Rect(Vector2(), quadTree.GetSize());
+
+    if (!bounds.Contains(start) || !bounds.Contains(dest) ||
+        !IsValidPosition(start, Vector2(__FLT_EPSILON__), quadTree,
+                         checkedOrders) ||
+        !IsValidPosition(dest, Vector2(__FLT_EPSILON__), quadTree,
+                         checkedOrders)) {
+        return std::deque<Vector2>();
+    }
+
+    const size_t width =
+        static_cast<size_t>((bounds.Width() / detailLevel) + 1);
+    const size_t height =
+        static_cast<size_t>((bounds.Height() / detailLevel) + 1);
+    // world grid
+    Node *grid = new Node[width * height];
+
+    const size_t startIndex =
+        ConvertVectorToNodeIndex(start, width, detailLevel);
+    const size_t endIndex = ConvertVectorToNodeIndex(dest, width, detailLevel);
+
+    std::set<size_t, PathFinding::Comparator> queue =
+        std::set<size_t, PathFinding::Comparator>(
+            PathFinding::Comparator(grid));
+    queue.insert(startIndex);
+    grid[startIndex].visited = true;
+    grid[startIndex].path = NULLVALUE;
+
+    size_t neighbors[8];
+    while (!queue.empty()) {
+        const size_t nodeIndex = *queue.begin();
+        queue.erase(queue.begin());
```

Comment:
`queue.pop()` would be nicer

Reply:
Yes, though std::set does not support that.